File: Python/Trees/trie.py
# Path: C/Trees/trie.c
# C program to implement trie.

# A prefix trie is an ordered tree data structure used in the representation of a set of strings over a finite alphabet set,
# which allows efficient storage of words with common prefixes.

class TrieNode:
    # Trie node
    def __init__(self):
        # 26 lowercase alphabets
        self.children = [None] * 26
        self.isEndOfWord = False


# Creates a new Trie node
def createNode():
    return TrieNode()


# Inserts a word into the Trie
def insert(root, word):
    currentNode = root

    for char in word:
        ind = ord(char) - ord('a')

        if currentNode.children[ind] is None:
            currentNode.children[ind] = createNode()

        currentNode = currentNode.children[ind]

    currentNode.isEndOfWord = True


# Searches for a word in the Trie
def search(root, word):
    currentNode = root

    for char in word:
        ind = ord(char) - ord('a')

        if currentNode.children[ind] is None:
            return False

        currentNode = currentNode.children[ind]

    return currentNode is not None and currentNode.isEndOfWord


# Checks if a Trie node has any children
def hasChildren(node):
    for child in node.children:
        if child is not None:
            return True

    return False


# Deletes a word from the Trie
def deleteWord(root, word, level):
    if root is None:
        return False

    if level == len(word):
        if root.isEndOfWord:
            root.isEndOfWord = False

            if not hasChildren(root):
                root = None

            return True

        return False

    ind = ord(word[level]) - ord('a')
    deleted = deleteWord(root.children[ind], word, level + 1)

    if deleted and not hasChildren(root.children[ind]):
        root.children[ind] = None

    return deleted


# Deletes the entire Trie
def deleteTrie(root):
    if root is None:
        return

    for child in root.children:
        if child is not None:
            deleteTrie(child)

    del root


# Program starts here
if __name__ == "__main__":
    root = createNode()

    # Inserting words into the Trie
    insert(root, "apple")
    insert(root, "banana")
    insert(root, "car")
    insert(root, "cat")
    insert(root, "cart")
    insert(root, "candid")
    print("Words: apple, banana, car, cat, cart, candid are Inserted.\n")

    print("Search results:")

    # Searching for words in the Trie
    print("apple:", "Found" if search(root, "apple") else "Not found")
    print("banana:", "Found" if search(root, "banana") else "Not found")
    print("care:", "Found" if search(root, "care") else "Not found")
    print("cat:", "Found" if search(root, "cat") else "Not found")
    print("cart:", "Found" if search(root, "cart") else "Not found")
    print("dog:", "Found" if search(root, "dog") else "Not found")

    # Deleting words from the Trie
    deleteWord(root, "banana", 0)

    print("\nAfter deleting 'banana':")
    print("banana:", "Found" if search(root, "banana") else "Not found")

    deleteWord(root, "car", 0)

    print("\nAfter deleting 'car':")
    print("car:", "Found" if search(root, "car") else "Not found")
    print("cart:", "Found" if search(root, "cart") else "Not found")

    # Deleting the entire Trie
    deleteTrie(root)

File: C/Trees/trie.c
// Path: C/Trees/trie.c
// C program to implement trie.

// A prefix trie is an ordered tree data structure used in the representation of a set of strings over a finite alphabet set,
// which allows efficient storage of words with common prefixes.

#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>

// Trie node
struct TrieNode
{
    // 26 lowercase alphabets
    struct TrieNode *children[26];
    bool isEndOfWord;
};

// Creates a new Trie node
struct TrieNode *createNode()
{
    struct TrieNode *newNode = (struct TrieNode *)malloc(sizeof(struct TrieNode));
    newNode->isEndOfWord = false;

    for (int i = 0; i < 26; i++)
    {
        newNode->children[i] = NULL;
    }

    return newNode;
}

// Inserts a word into the Trie
void insert(struct TrieNode *root, const char *word)
{
    struct TrieNode *currentNode = root;

    for (int i = 0; word[i] != '\0'; i++)
    {
        int ind = word[i] - 'a';

        if (currentNode->children[ind] == NULL)
        {
            currentNode->children[ind] = createNode();
        }

        currentNode = currentNode->children[ind];
    }

    currentNode->isEndOfWord = true;
}

// Searches for a word in the Trie
bool search(struct TrieNode *root, const char *word)
{
    struct TrieNode *currentNode = root;

    for (int i = 0; word[i] != '\0'; i++)
    {
        int ind = word[i] - 'a';

        if (currentNode->children[ind] == NULL)
        {
            return false;
        }

        currentNode = currentNode->children[ind];
    }

    return (currentNode != NULL && currentNode->isEndOfWord);
}

// Checks if a Trie node has any children
bool hasChildren(struct TrieNode *node)
{
    for (int i = 0; i < 26; i++)
    {
        if (node->children[i] != NULL)
        {
            return true;
        }
    }

    return false;
}

// Deletes a word from the Trie
bool deleteWord(struct TrieNode *root, const char *word, int level)
{
    if (root == NULL)
    {
        return false;
    }

    if (word[level] == '\0')
    {
        if (root->isEndOfWord)
        {
            root->isEndOfWord = false;

            if (!hasChildren(root))
            {
                free(root);
                root = NULL;
            }

            return true;
        }

        return false;
    }

    int ind = word[level] - 'a';
    bool deleted = deleteWord(root->children[ind], word, level + 1);

    if (deleted && !hasChildren(root->children[ind]))
    {
        free(root->children[ind]);
        root->children[ind] = NULL;
    }

    return deleted;
}

// Deletes the entire Trie
void deleteTrie(struct TrieNode *root)
{
    if (root == NULL)
    {
        return;
    }

    for (int i = 0; i < 26; i++)
    {
        if (root->children[i] != NULL)
        {
            deleteTrie(root->children[i]);
        }
    }

    free(root);
}

// Program starts here
int main()
{
    struct TrieNode *root = createNode();

    // Inserting words into the Trie
    insert(root, "apple");
    insert(root, "banana");
    insert(root, "car");
    insert(root, "cat");
    insert(root, "cart");
    insert(root, "candid");
    printf("Words: apple, banana, car, cat, cart, candid are Inserted.\n\n");

    printf("Search results:\n");

    // Searching for words in the Trie
    printf("apple: %s\n", search(root, "apple") ? "Found" : "Not found");
    printf("banana: %s\n", search(root, "banana") ? "Found" : "Not found");
    printf("care: %s\n", search(root, "care") ? "Found" : "Not found");
    printf("cat: %s\n", search(root, "cat") ? "Found" : "Not found");
    printf("cart: %s\n", search(root, "cart") ? "Found" : "Not found");
    printf("dog: %s\n", search(root, "dog") ? "Found" : "Not found");

    // Deleting words from the Trie
    deleteWord(root, "banana", 0);

    printf("\nAfter deleting 'banana':\n");
    printf("banana: %s\n", search(root, "banana") ? "Found" : "Not found");

    deleteWord(root, "car", 0);

    printf("\nAfter deleting 'car':\n");
    printf("car: %s\n", search(root, "car") ? "Found" : "Not found");
    printf("cart: %s\n", search(root, "cart") ? "Found" : "Not found");

    // Deleting the entire Trie
    deleteTrie(root);

    return 0;
}
